pkg/operator/controller/status/controller.go

```diff
@@ -864,3 +984,38 @@ func joinConditions(currCond, newCond configv1.ClusterOperatorStatusCondition) c
 	}
 	return currCond
 }
+
+// compareVersionNums compares two strings of format "<name>.v#.#.#". It returns a negative number if 'a' has a higher
+// version number, a positive number if 'b' has a higher version number, or 0 if the version numbers are identical. If
+// it is unable to parse either version number, or if the names from the two version strings differ, an error is
+// returned and the comparison should be considered invalid.
+func compareVersionNums(a, b string) (int, error) {
+	var aName, bName string
+	var aX, aY, aZ, bX, bY, bZ int
+	aSplit := strings.Split(a, ".")
+	if len(aSplit) != 4 {
+		return 0, fmt.Errorf("%q does not match expected format", a)
+	}
+	aName = aSplit[0]
+	aX, _ = strconv.Atoi(aSplit[1][1:]) // X has format "v%d", so [1:] cuts out the "v"
+	aY, _ = strconv.Atoi(aSplit[2])
+	aZ, _ = strconv.Atoi(aSplit[3])
+	bSplit := strings.Split(b, ".")
+	if len(bSplit) != 4 {
+		return 0, fmt.Errorf("%q does not match expected format", b)
+	}
+	bName = bSplit[0]
+	bX, _ = strconv.Atoi(bSplit[1][1:]) // X has format "v%d", so [1:] cuts out the "v"
+	bY, _ = strconv.Atoi(bSplit[2])
+	bZ, _ = strconv.Atoi(bSplit[3])
+	if aName != bName {
+		return 0, fmt.Errorf("%q and %q are different packages. cannot compare version numbers", a, b)
+	}
```

Comment:
It would be better to check for the same package name before we even call this function.

Reply:
This is more of a safety check. We shouldn't hit this in actual use.

Reply:
Yeah. Currently, `compareVersionNums` only used in the case where `subscription.Spec.Package` is `servicemeshoperator3`, and it's compared against the expected version, which will be `servicemeshoperator3.v<version>`. It's possible someone could change the expected version, since it's a field set in the ingress operator deployment, but that's not really supported.